client/exec.go

```diff
@@ -0,0 +1,312 @@
+		Height:        opts.Height,
+	}
+	var body bytes.Buffer
+	err := json.NewEncoder(&body).Encode(&payload)
+	if err != nil {
+		return nil, err
+	}
+	headers := map[string]string{
+		"Content-Type": "application/json",
+	}
+	resultBytes, changeID, err := client.doAsyncFull("POST", "/v1/exec", nil, headers, &body)
+	if err != nil {
+		return nil, err
+	}
+	var result execResult
+	err = json.Unmarshal(resultBytes, &result)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check that we have the websocket IDs we expect.
+	wsIDs := result.WebsocketIDs
+	if wsIDs["control"] == "" {
+		return nil, fmt.Errorf(`response did not include "control" websocket`)
+	}
+	if wsIDs["io"] == "" {
+		return nil, fmt.Errorf(`response did not include "io" websocket`)
+	}
+	if !opts.CombineStderr && wsIDs["stderr"] == "" {
+		return nil, fmt.Errorf(`response did not include "stderr" websocket`)
+	}
+
+	// Connect to the "control" websocket.
+	controlConn, err := client.getChangeWebsocket(changeID, wsIDs["control"])
+	if err != nil {
+		return nil, err
+	}
+
+	// Forward stdin and stdout.
+	ioConn, err := client.getChangeWebsocket(changeID, wsIDs["io"])
+	if err != nil {
+		return nil, err
+	}
+	stdinDone := wsutil.WebsocketSendStream(ioConn, stdin, -1)
+	stdoutDone := wsutil.WebsocketRecvStream(stdout, ioConn)
+
+	// Handle stderr separately if needed.
+	var stderrConn *websocket.Conn
+	var stderrDone chan bool
+	if !opts.CombineStderr {
+		stderrConn, err = client.getChangeWebsocket(changeID, wsIDs["stderr"])
+		if err != nil {
+			return nil, err
+		}
+		stderrDone = wsutil.WebsocketRecvStream(stderr, stderrConn)
+	}
+
+	// Fire up a goroutine to wait for everything to be done.
+	writesDone := make(chan struct{})
+	go func() {
+		// Wait till the WebsocketRecvStream goroutines are done writing to
+		// stdout and stderr. This happens when EOF is signalled or websocket
+		// is closed.
+		<-stdoutDone
+		if stderrDone != nil {
+			<-stderrDone
+		}
+
+		// Empty the stdin channel, but don't block on it as stdin may be
+		// stuck in Read. This is due to the somewhat poor design of
+		// WebsocketSendStream, which writes to an unbuffered channel
+		// (instead of closing it or using a buffered channel). We'd rather
+		// not modify that package much, so it's closer to the LXD code.
+		go func() {
+			<-stdinDone // happens when reading stdin returns EOF
+		}()
+
+		// Try to close websocket connections gracefully, but ignore errors.
+		_ = ioConn.Close()
+		if stderrConn != nil {
+			_ = stderrConn.Close()
+		}
+		_ = controlConn.Close()
+
+		// Tell ExecProcess.Wait we're done writing to stdout/stderr.
+		close(writesDone)
+	}()
+
+	process := &ExecProcess{
+		changeID:    changeID,
+		client:      client,
+		timeout:     opts.Timeout,
+		writesDone:  writesDone,
+		controlConn: controlConn,
+	}
+	return process, nil
+}
+
+// Wait waits for the command process to finish and returns its exit code.
+func (p *ExecProcess) Wait() (int, error) {
```

Comment:
Should we return a type here instead of the bare exit code, so we can reserve the option of returning more details about the termination?  Alternatively, we might also make it return only an error, and add an additional method on ExecProcess that returns the termination information, with the first one being `ExitCode`?

Reply:
Good thinking. I ended up changing it to be similar to `os/exec`'s `Cmd.Wait()`: it returns just an error, which is non-nil if there's any error or if the process exits with a non-zero exit code. This makes it easier for the caller to check for an error in any case -- there's no need to check the error return *and* the exit code (which you might forget to do).

`Cmd.Wait` does this too. For an exit with a non-zero exit code, it returns an error of type `*exec.ExitError`, which has an `ExitCode` method. I've mirrored that, wait `Wait()` returning an error of type `*client.ExitError` with an `ExitCode` method. It's definitely easier to use, as the removal of lines in the tests show.

Reply:
Sounds good!